### PR TITLE
batch creation and list vm support

### PR DESCRIPTION
**Pull request by yb01**
Create VMs in batch with replicaset, with reservation_id=true
List VM with the reservation_id

Local test:
```
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# kubectl get pods -AT
TENANT   NAMESPACE     NAME                              HASHKEY               READY   STATUS    RESTARTS   AGE
system   kube-system   coredns-default-649bd7994-rf4qm   8387358657457246339   1/1     Running   0          17m
system   kube-system   kube-dns-554c5866fc-cgcs2         1553543069423076189   3/3     Running   0          17m
system   kube-system   virtlet-jl6js                     2980399683183508053   3/3     Running   0          17m
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# curl -L -k -XPOST -H "Content-Type: application/json" -H "User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/$Format" -H "Accept: application/json" -H "openstack: true" 'http://localhost:8080/servers' -d @test/e2e/arktos/testvm-batch.json -v | jq
Note: Unnecessary use of -X or --request, POST is already inferred.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /servers HTTP/1.1
> Host: localhost:8080
> Content-Type: application/json
> User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/
> Accept: application/json
> openstack: true
> Content-Length: 244
> 
} [244 bytes data]
* upload completely sent off: 244 out of 244 bytes
< HTTP/1.1 307 Temporary Redirect
< Cache-Control: no-cache, private
< Location: /apis/apps/v1/tenants/system/namespaces/kube-system/replicasets
< Date: Wed, 19 Jan 2022 02:38:25 GMT
< Content-Length: 0
< 
100   244    0     0  100   244      0   238k --:--:-- --:--:-- --:--:--  238k
* Connection #0 to host localhost left intact
* Issue another request to this URL: 'http://localhost:8080/apis/apps/v1/tenants/system/namespaces/kube-system/replicasets'
* Found bundle for host localhost: 0x55cd414bb680 [can pipeline]
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (127.0.0.1) port 8080 (#0)
> POST /apis/apps/v1/tenants/system/namespaces/kube-system/replicasets HTTP/1.1
> Host: localhost:8080
> Content-Type: application/json
> User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/
> Accept: application/json
> openstack: true
> Content-Length: 244
> 
} [244 bytes data]
* upload completely sent off: 244 out of 244 bytes
< HTTP/1.1 201 Created
< Cache-Control: no-cache, private
< Content-Type: application/json
< Date: Wed, 19 Jan 2022 02:38:25 GMT
< Content-Length: 28
< 
{ [28 bytes data]
100   272  100    28  100   244   5600  48800 --:--:-- --:--:-- --:--:-- 54400
* Connection #0 to host localhost left intact
{
  "reservation_id": "testvm"
}
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# curl -L -k -XGET -H "Content-Type: application/json" -H "User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/$Format" -H "Accept: application/json" -H "openstack: true" 'http://localhost:8080/servers' -d @test/e2e/arktos/testvm-list.json -v | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /servers HTTP/1.1
> Host: localhost:8080
> Content-Type: application/json
> User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/
> Accept: application/json
> openstack: true
> Content-Length: 29
> 
} [29 bytes data]
* upload completely sent off: 29 out of 29 bytes
< HTTP/1.1 307 Temporary Redirect
< Cache-Control: no-cache, private
< Content-Type: text/html; charset=utf-8
< Location: /api/v1/tenants/system/namespaces/kube-system/pods?labelSelector=ln=testvm
< Date: Wed, 19 Jan 2022 02:38:44 GMT
< Content-Length: 110
< 
* Ignoring the response-body
{ [110 bytes data]
100   139  100   110  100    29  15714   4142 --:--:-- --:--:-- --:--:-- 19857
* Connection #0 to host localhost left intact
* Issue another request to this URL: 'http://localhost:8080/api/v1/tenants/system/namespaces/kube-system/pods?labelSelector=ln=testvm'
* Found bundle for host localhost: 0x5601c62ddae0 [can pipeline]
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /api/v1/tenants/system/namespaces/kube-system/pods?labelSelector=ln=testvm HTTP/1.1
> Host: localhost:8080
> Content-Type: application/json
> User-Agent: kubectl/v0.9.0 (linux/amd64) kubernetes/
> Accept: application/json
> openstack: true
> Content-Length: 29
> 
} [29 bytes data]
* upload completely sent off: 29 out of 29 bytes
< HTTP/1.1 200 OK
< Cache-Control: no-cache, private
< Content-Type: application/json
< Date: Wed, 19 Jan 2022 02:38:44 GMT
< Content-Length: 386
< 
{ [386 bytes data]
100   415  100   386  100    29  29692   2230 --:--:-- --:--:-- --:--:-- 31922
* Connection #0 to host localhost left intact
{
  "Servers": [
    {
      "Id": "testvm-dl84b",
      "Links": [
        {
          "Link": "/api/v1/namespaces/kube-system/pods/testvm-dl84b",
          "Rel": ""
        }
      ],
      "Security_groups": null
    },
    {
      "Id": "testvm-wcg4z",
      "Links": [
        {
          "Link": "/api/v1/namespaces/kube-system/pods/testvm-wcg4z",
          "Rel": ""
        }
      ],
      "Security_groups": null
    },
    {
      "Id": "testvm-xfmhp",
      "Links": [
        {
          "Link": "/api/v1/namespaces/kube-system/pods/testvm-xfmhp",
          "Rel": ""
        }
      ],
      "Security_groups": null
    }
  ]
}
root@ip-172-31-10-115:/work/src/k8s.io/kubernetes# 

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go [hunk 1]
```diff
@@ -110,21 +162,73 @@ func (o Openstack) genericOpenStackRequestHandler(target string, resp http.Respo
...
+			redirectUrl += "/" + getElementFromPath(req.URL.Path)
+		}
+
+		rev_id, err := peekListFilters(req)
+		if err != nil {
+			resp.WriteHeader(http.StatusBadRequest)
+			return
+		}
+		if rev_id != "" {
+			redirectUrl += fmt.Sprintf("?labelSelector=ln=%s", rev_id)
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:212]: This looks weird "labelSelector=ln=". Have you tested it?
If it works, can you add a sample URL as a comment?

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go [hunk 2]
```diff
@@ -110,21 +162,73 @@ func (o Openstack) genericOpenStackRequestHandler(target string, resp http.Respo
...
+		}
+
+	case http.MethodDelete:
+		redirectUrl = fmt.Sprintf(POD_URL_TEMPLATE, tenant, namespace)
 		redirectUrl += "/" + getElementFromPath(req.URL.Path)
+	case http.MethodPost:
+		batchRequest, err := peekBatchRequest(req)
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:219]: rename to isBacthRequest

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go [hunk 3]
```diff
@@ -323,3 +401,7 @@ func GetNamespaceFromRequest(r *http.Request) string {
 func IsActionRequest(path string) bool {
 	return strings.HasSuffix(path, "action")
 }
+
+func IsBatchCreationRequest(r OpenstackServerRequest) bool {
+	return r.Min_count > 0
```
Comment (Sindica) [inline staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:406]: >1?

File: staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go [hunk 4]
```diff
@@ -206,7 +206,7 @@ func (c *codec) doEncode(obj runtime.Object, w io.Writer) error {
 	// Openstack is not Arktos API object
 	// simply encode it and return
 	typeStr := reflect.TypeOf(obj).String()
-	if typeStr == "*openstack.OpenstackRebuildResponse" || typeStr == "*openstack.OpenstackCreateImageResponse" || typeStr == "*openstack.OpenstackResponse" {
+	if typeStr == "*openstack.OpenstackServerListResponse" || typeStr == "*openstack.OpenstackBatchResponse" || typeStr == "*openstack.OpenstackRebuildResponse" || typeStr == "*openstack.OpenstackCreateImageResponse" || typeStr == "*openstack.OpenstackResponse" {
```
[the remark this answers is not in the document]
Comment (yb01) [reply staging/src/k8s.io/apimachinery/pkg/runtime/serializer/versioning/versioning.go:209]: No, per Openstack api doc,  multiple server(groups) only has CREATE, and LIST for servers under same "reservation_id"

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go [hunk 5]
```diff
@@ -45,6 +48,55 @@ func getElementFromPath(path string) string {
...
 }
 
+func isGetDetail(path string) bool {
+	pattern := "/servers/[a-z]+"
+	match, _ := regexp.MatchString(pattern, path)
+
+	return match
+}
+
+func peekBatchRequest(req *http.Request) (bool, error) {
```
[the remark this answers is not in the document]
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:58]: sure will change to match the returns of the func

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go [hunk 6]
```diff
@@ -45,6 +48,55 @@ func getElementFromPath(path string) string {
...
+	if err != nil {
+		klog.Errorf("error unmarshal request. error %v", err)
+		return false, err
+	}
+
+	return openstack.IsBatchCreationRequest(obj), nil
+}
+
+// For now, supported filter is the reservation_id when VMs were created in batch
+func peekListFilters(req *http.Request) (string, error) {
```
[the remark this answers is not in the document]
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:78]: sure will change to match the returns of the func

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go (hunk 1 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:212]: yes, this is tested as described in the PR.

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go (hunk 2 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go:219]: sure, will change.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go (hunk 3 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:406]: no. 1 will be considered as batch request as well. i.e. a group with one server.

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go (hunk 2 above)
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go]: fixed.

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go]: changed.

File: staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/server/routes/openstack.go]: fixed.

File: staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go (hunk 3 above)
Comment (Sindica) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:406]: The real question is what will be the Min_count value for a non batch request?

Comment (Sindica) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:406]: Please add some comment so that there will be no further similar questions.

Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:406]: 0, which is the default golang int value will be no-batch cases -- note that in the creation request,  for no-batch case the Openstack request from user ( the json file ) has no min_count, max_count per Openstack spec.  internally we use the same data structure.
the comment are already there in the OpenstackServerRequest{}. i will add some here too.

Comment (yb01) [reply staging/src/k8s.io/apiserver/pkg/util/openstack/openstackutils.go:410]: updated.

**Review by Sindica — APPROVED** [review]
/lgtm

**Pull request merged**